h2/src/main/org/h2/command/dml/Select.java

```diff
@@ -569,7 +569,8 @@ protected ResultInterface queryWithoutCache(int maxRows, ResultTarget target) {
         }
         boolean lazy = session.isLazyQueryExecution() &&
                 target == null && !isForUpdate && !isQuickAggregateQuery &&
-                limitRows != 0 && offsetExpr == null && isReadOnly();
+                limitRows != 0 && offsetExpr == null && isReadOnly() && 
+                !hasRecursiveTableView();
```

Comment:
The recursive CTE itself can not be lazily executed, but why do you disable lazy execution for the outer query? I do not think it should be like this.

Reply:
@svladykin Thanks for commenting - do you have a suggestion of what level to put this check against ? By all means I'll try it out. This was the first attempt at fixing it, that worked....

And I am no expert on lazy, myself.

Reply:
@svladykin I am trying a different condition for excluding lazy eval  - just looking at the topTableFilter - I hope this is what you were asking for...

Reply:
Thank you for working on this issue, but what is the reasoning for checking only the topTableFilter? I mean it looks like you are trying to do some random changes and hope that the tests will pass. Obviously, this is not how high-quality software is being made. Maybe we just do not have enough test coverage to break your code.

Let's try to figure out what happens if we remove this condition altogether. Why do the tests start failing? To my knowledge even if the topTableFilter is recursive, the recursive query must be executed actively, produce an intermediate result set that can be lazily joined to subsequent table filters. Why it does not work like this? Can you make this change and provide a list of failing tests so that I can take a look?

Reply:
This was the failure without this "fix" - it seems it was not just a lazy eval issue - but a combination of things that triggered it ( I wasn't able to reproduce it as a stand alone test case):

Test lazy mvStore memory mvcc multiThreaded  (20759 KB used)
01:37:12 00:00.000 org.h2.test.scripts.TestScriptSimple 
01:37:12 00:00.000 org.h2.test.scripts.TestScript Running commands in testScript.sql
ERROR: Unexpected General error select sum(n) from (
with recursive r(n) as (
(select 1) union all (select n+1 from r where n < 3)
)
select n from r
) org.h2.jdbc.JdbcSQLException: General error: "java.lang.StackOverflowError"; SQL statement:
select sum(n) from (
with recursive r(n) as (
(select 1) union all (select n+1 from r where n < 3)
)
select n from r
) [50000-196] ------------------------------
org.h2.jdbc.JdbcSQLException: General error: "java.lang.StackOverflowError"; SQL statement:
select sum(n) from (
with recursive r(n) as (
(select 1) union all (select n+1 from r where n < 3)
)
select n from r
) [50000-196]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:168)
	at org.h2.message.DbException.convert(DbException.java:291)
	at org.h2.command.Command.executeQuery(Command.java:215)
	at org.h2.jdbc.JdbcStatement.executeInternal(JdbcStatement.java:186)
	at org.h2.jdbc.JdbcStatement.execute(JdbcStatement.java:164)
	at org.h2.test.scripts.TestScript.processStatement(TestScript.java:318)
	at org.h2.test.scripts.TestScript.process(TestScript.java:246)
	at org.h2.test.scripts.TestScript.testFile(TestScript.java:208)
	at org.h2.test.scripts.TestScript.testScript(TestScript.java:160)
	at org.h2.test.scripts.TestScript.test(TestScript.java:81)
	at org.h2.test.TestBase.runTest(TestBase.java:145)
	at org.h2.test.TestAll.addTest(TestAll.java:987)
	at org.h2.test.TestAll.test(TestAll.java:724)
	at org.h2.test.TestAll.runTests(TestAll.java:636)
	at org.h2.test.TestAll.testAll(TestAll.java:566)
	at org.h2.test.TestAll.run(TestAll.java:515)
	at org.h2.test.TestAll.main(TestAll.java:455)
Caused by: java.lang.StackOverflowError
	at org.h2.command.dml.Select$LazyResultSelect.reset(Select.java:1452)
	at org.h2.index.ViewIndex.findRecursive(ViewIndex.java:188)
	at org.h2.index.ViewIndex.find(ViewIndex.java:285)
	at org.h2.index.ViewIndex.find(ViewIndex.java:161)
	at org.h2.index.BaseIndex.find(BaseIndex.java:127)
	at org.h2.index.IndexCursor.find(IndexCursor.java:169)
	at org.h2.table.TableFilter.next(TableFilter.java:468)
	at org.h2.command.dml.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1475)
	at org.h2.result.LazyResult.hasNext(LazyResult.java:79)
	at org.h2.result.LazyResult.next(LazyResult.java:59)
	at org.h2.index.ViewCursor.next(ViewCursor.java:49)
	at org.h2.index.IndexCursor.next(IndexCursor.java:309)
	at org.h2.table.TableFilter.next(TableFilter.java:499)
	at org.h2.command.dml.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1475)
	at org.h2.result.LazyResult.hasNext(LazyResult.java:79)
	at org.h2.result.LazyResult.next(LazyResult.java:59)
	at org.h2.index.ViewCursor.next(ViewCursor.java:49)
	at org.h2.index.IndexCursor.next(IndexCursor.java:309)
	at org.h2.table.TableFilter.next(TableFilter.java:499)
	at org.h2.command.dml.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1475)
	at org.h2.result.LazyResult.hasNext(LazyResult.java:79)
	at org.h2.result.LazyResult.next(LazyResult.java:59)
	at org.h2.index.ViewCursor.next(ViewCursor.java:49)
	at org.h2.index.IndexCursor.next(IndexCursor.java:309)
	at org.h2.table.TableFilter.next(TableFilter.java:499)
	at org.h2.command.dml.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1475)
	at org.h2.result.LazyResult.hasNext(LazyResult.java:79)
	at org.h2.result.LazyResult.next(LazyResult.java:59)
	at org.h2.index.ViewCursor.next(ViewCursor.java:49)
	at org.h2.index.IndexCursor.next(IndexCursor.java:309)
	at org.h2.table.TableFilter.next(TableFilter.java:499)
	at org.h2.command.dml.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1475)
<<<<<<<<<< ... stack trace pattern continues as above... >>>>>>>>>>>>

Reply:
Looks weird. Probably for some reason the recursive CTE is executed lazily while it should not. 

BTW, it should not be a problem to reproduce it, you need to just manually set all the needed flags in TestScript and run just it.